tests/epyccel/recognised_functions/test_numpy_funcs.py

```diff
@@ -5910,46 +5658,522 @@ def get_prod(arr):
+    integer16 = randint(min_int16, max_int16, size=size, dtype=np.int16)
+    integer = randint(min_int, max_int, size=size, dtype=np.int)
+    integer32 = randint(min_int32, max_int32, size=size, dtype=np.int32)
+    integer64 = randint(min_int64, max_int64, size=size, dtype=np.int64)
+
+    fl = uniform(min_int, max_int, size = size)
+    fl32 = uniform(min_int, max_int, size = size)
+    fl32 = np.float32(fl32)
+    fl64 = uniform(min_int64, max_int64, size = size)
+
+    cmplx128_from_float32 = uniform(low=min_float32 / 2, high=max_float32 / 2, size=size) + uniform(low=min_float32 / 2, high=max_float32 / 2, size=size) * 1j
+    # the result of the last operation is a Python complex type which has 8 bytes in the alignment,
+    # that's why we need to convert it to a numpy.complex64 the needed type.
+    cmplx64 = np.complex64(cmplx128_from_float32)
+    cmplx128 = uniform(low=min_float64 / 2, high=max_float64 / 2, size=size) + uniform(low=min_float64 / 2, high=max_float64 / 2, size=size) * 1j
+
+    epyccel_func = epyccel(get_matmul, language=language)
+
+    assert epyccel_func(bl) == get_matmul(bl)
+    assert epyccel_func(integer8) == get_matmul(integer8)
+    assert epyccel_func(integer16) == get_matmul(integer16)
+    assert epyccel_func(integer) == get_matmul(integer)
+    assert epyccel_func(integer32) == get_matmul(integer32)
+    # the if block should be removed after resolving (https://github.com/pyccel/pyccel/issues/735).
+    if sys.platform != 'win32':
+        assert epyccel_func(integer64) == get_matmul(integer64)
+    assert epyccel_func(fl) == get_matmul(fl)
+    assert epyccel_func(fl32) == get_matmul(fl32)
+    assert epyccel_func(fl64) == get_matmul(fl64)
+    assert (epyccel_func(cmplx64) == get_matmul(cmplx64))
+    assert (epyccel_func(cmplx128) == get_matmul(cmplx128))
+
+@pytest.mark.parametrize( 'language', (
+        pytest.param("fortran", marks = [pytest.mark.fortran]),
+       #     pytest.mark.skip(reason="TypeError: cannot determine the type of <pyccel.ast.numpyext.NumpyMatmul object at 0x1218b9d60> (semantic.py -> numpyext.py)")]),
+        pytest.param("c", marks = [
+            pytest.mark.skip(reason="Needs a C printer see https://github.com/pyccel/pyccel/issues/791"),
+            pytest.mark.c]
+        ),
+        pytest.param("python", marks = [
+            pytest.mark.python],
+        #    pytest.mark.skip(reason="Missing Python printer")]
+        )
+    )
+)
+
+def test_numpy_matmul_array_like_2x2d(language):
+
+    @types('bool[:,:]')
+    @types('int[:,:]')
+    @types('int8[:,:]')
+    @types('int16[:,:]')
+    @types('int32[:,:]')
+    @types('int64[:,:]')
+    @types('float[:,:]')
+    @types('float32[:,:]')
+    @types('float64[:,:]')
+    @types('complex64[:,:]')
+    @types('complex128[:,:]')
+    def get_matmul(arr):
+        from numpy import matmul, shape
+        a = matmul(arr, arr)
+        s = shape(a)
+        return len(s) , s[0] , s[1] , a[0,1] , a[1,0]
```

Comment:
`a[0,1]` and `a[1,0]` are int64 if `arr` is `int64[:]` which means they are affected by issue #735